research/quantifying-data-leakage/README.md

```diff
@@ -1,16 +1,260 @@
+prepared dataset list for client 9. 
+```
+python3 -m nvflare_gradinv.utils.modify_config --config_file \
+./job_configs/cxr_1client/client/config/config_fed_client.json \
+--data_root=${DATA_ROOT} \
+--dataset_json=${PWD}/data/data_200val+rest_client9.json
+```
+
+#### cxr_9clients
+To run a full FL experiment, we prepared several clients with different 
+dataset sizes and batch sizes.
+
+| Client                 | Batch Size | # Training | # Validation |  
+|------------------------|------------|------------|--------------|
+| client 1               | 4          | 8          | 200          |
+| client 2               | 4          | 32         | 200          |
+| client 3               | 4          | 128        | 200          |
+| client 4               | 4          | 512        | 200          |
+| client 5               | 8          | 8          | 200          |
+| client 6               | 8          | 32         | 200          |
+| client 7               | 8          | 128        | 200          |
+| client 8               | 8          | 512        | 200          |
+| client 9 ("high-risk") | 1          | 1          | 200          |
+| Sum                    |            | 1360       | 1600         |
+Nr. Testing: 1382
+
+We can modify to client config files to use the downloaded data and above 
+dataset using
+```
+for i in {1..9}
+do
+  python3 -m nvflare_gradinv.utils.modify_config --config_file \
+  ./job_configs/cxr_9clients/client${i}/config/config_fed_client.json \
+  --data_root=${DATA_ROOT} \
+  --dataset_json=${PWD}/data/data_200val+rest_client${i}.json
+done
+```
+
+
+### 2. Start FL experiment
+
+To run the "quick" FL experiment, inverting one client, execute
+```
+N_CLIENTS=1
+MMAR=./job_configs/cxr_1client
+EXP_NAME=exp1
+./start_fl_sim.sh ${N_CLIENTS} ${MMAR} ${EXP_NAME}
+```
+
+To run the full FL experiment, inverting 3 out of 9 client, execute
+```
+N_CLIENTS=9
+MMAR=./job_configs/cxr_9clients
+EXP_NAME=exp1
+./start_fl_sim.sh ${N_CLIENTS} ${MMAR} ${EXP_NAME}
+```
+
+### 3. Visualize results
+In order to reduce the possibility for data leakage through gradient 
+inversion attacks, we use the [GaussianPrivacy](./src/nvflare_gradinv/filters/gaussian_privacy.py) filter.
+The filter's `sigma0=` argument defines how much noise is to be added to 
+the outgoing model updates. In order to compute the added noise adaptively 
+with respect to the model updates, we compute the 95th 
+percentile of absolute model update values and multiply this value by `sigma0` to arrive 
+at the `sigma` used to define Gaussian distribution with zero mean. 
+This Gaussian distribution is then sampled to add calibrated noise to each 
+model update.
+
+#### Inverted training images
+As an example, we show the reconstruction from **site-9** sending updates 
+based only one image (which is "high-risk") with different levels of `sigma0`. 
+Sending updates based on a few number of images and low batch sizes can 
+leak a significant amount of information. 
```

Comment:
(, hence client-9 is an extreme case with 1 image and batch size 1) - shall we move this part to earlier section give a general idea / description of the basic ideas and experiment settings?

Reply:
good point. it is explained also earlier already.